### PR TITLE
Fix output lifetime collection in needless_lifetimes

### PR DESCRIPTION
**Pull request by m-rph**
This is currently draft as I am unsure about how to go with fixing this. I don't completely understand the code D:, but I've identified the issue more or less.

## The problem

In the following snippet
```rust
mod issue11291 {
    use std::collections::HashMap;
    pub struct MyContainer(HashMap<u8, u32>);
    impl MyContainer {
        pub fn iter<'a>(&'a self) -> impl Iterator<Item = (&'a u8, &'a u32)> + 'a {
            //~^ ERROR: the following explicit lifetimes could be elided: 'a
            self.0.iter()
        }
    }
}
```
Clippy will suggest:

```
 pub fn iter<'a>(&'a self) -> impl Iterator<Item = (&'_ u8, &'a u32)> + 'a
```

Which misses the 2 latter lifetimes.

I probed a bit into the implementation. The lifetimes of the output are collected via a walker. The implementation of the walker is identical for both input and output declaration.

The problem seems related to the truncation in `TyKind::OpaqueDef`

https://github.com/rust-lang/rust-clippy/blob/86717f2f0c9630e51d8ab8fefb997db89d38ad8f/clippy_lints/src/lifetimes.rs#L532-L542

Without the truncation, all lifetimes are collected, but going by `hir::DefId`, we have 4 different lifetimes (LTs) instead of 3. Going by span, we have 3 LTs.

Removing the truncation and keeping only unique LT by their span causes something to break in the linting logic.

Update 1:

It seems that the issue is related to the defids of the LTs, as they can't be found in the allowed LTs.

Update 2:

Yup, using localdefid causes issues with opaque types. Using identifiers for any checks works better and the types are correlated selected. need to fix how the suggestions work to account for that, but otherwise it seems correct-ish. 

```
changelog: [`needless_lifetimes`]: Suggested fix will now elide all output lifetimes.
```

fixes: #12085  
fixes: #11291

r? @blyxyas

### CODE REVIEW

File: tests/ui/needless_lifetimes.stderr [hunk 1]
```diff
@@ -6,552 +6,297 @@ LL | fn distinct_lifetimes<'a, 'b>(_x: &'a u8, _y: &'b u8, _z: u8) {}
...
    |
 LL |     fn multiple_inputs_output_not_elided<'a, 'b>(x: &'a u8, y: &'b u8, z: &'b u8) -> &'b u8 {
-   |                                          ^^          ^^
-   |
-help: elide the lifetimes
-   |
-LL -     fn multiple_inputs_output_not_elided<'a, 'b>(x: &'a u8, y: &'b u8, z: &'b u8) -> &'b u8 {
-LL +     fn multiple_inputs_output_not_elided<'b>(x: &u8, y: &'b u8, z: &'b u8) -> &'b u8 {
-   |
+   |                                          ^^          ^^         ^^         ^^         ^^
```
Comment (m-rph) [inline tests/ui/needless_lifetimes.stderr:272]: This is wrong, it wrongly collects both `'a` and `'b`, should only elide `'a`

File: tests/ui/needless_lifetimes.stderr [hunk 2]
```diff
@@ -6,552 +6,297 @@ LL | fn distinct_lifetimes<'a, 'b>(_x: &'a u8, _y: &'b u8, _z: u8) {}
...
    |
 LL | fn alias_with_lt4a<'a, 'b>(_foo: &'a FooAlias<'b>) -> &'a str {
-   |                        ^^                     ^^
-   |
-help: elide the lifetimes
-   |
-LL - fn alias_with_lt4a<'a, 'b>(_foo: &'a FooAlias<'b>) -> &'a str {
-LL + fn alias_with_lt4a<'a>(_foo: &'a FooAlias<'_>) -> &'a str {
-   |
+   |                        ^^         ^^          ^^       ^^
```
Comment (m-rph) [inline tests/ui/needless_lifetimes.stderr:128]: `'a` shouldn't be collected.

File: tests/ui/needless_lifetimes.stderr [hunk 3]
```diff
@@ -6,552 +6,297 @@ LL | fn distinct_lifetimes<'a, 'b>(_x: &'a u8, _y: &'b u8, _z: u8) {}
...
    |
 LL |     fn bar<'a>(x: &'a u8, y: &'_ u8, z: &'_ u8) {}
-   |            ^^      ^^
-   |
-help: elide the lifetimes
-   |
-LL -     fn bar<'a>(x: &'a u8, y: &'_ u8, z: &'_ u8) {}
-LL +     fn bar(x: &u8, y: &'_ u8, z: &'_ u8) {}
-   |
+   |            ^^      ^^         ^^         ^^
```
Comment (m-rph) [inline tests/ui/needless_lifetimes.stderr:260]: Same issue as above.

**Review by m-rph — COMMENTED** [review]
Current commit results in a regression where more LTs are collected than they should be.

**Comment by xFrednet** [comment]
Hey @m-rph, this is a ping from triage, since there hasn't been any activity in some time. Are you still planning to continue this implementation?

@rustbot author

**Comment by m-rph** [comment]
@xFrednet yup. Need to find time. There's 5 issues that need to be addressed and will do all of them in 1 ~~rust~~ go.

**Comment by xFrednet** [comment]
Okay, just here to poke you a bit ^^ Let me know if you need any help.

```
error: found time
  --> https://github.com/rust-lang/rust-clippy/pull/12456/comment-5/1:29
   |
   | @xFrednet yup. Need to find time. 
   |                             ^^^^
   = note: This `time` might be a different type then the
           one you where refering to
```

**Comment by xFrednet** [comment]
The assignment was a test. I'm unassigning myself again. Thank you for the help!

**Comment by xFrednet** [comment]
Hey, @m-rph this is a ping from triage, since there hasn't been any activity in some time. Are you still planning to continue this implementation?

Also, is this PR maybe outdated, since you crated https://github.com/rust-lang/rust-clippy/pull/13141?

@rustbot author

**Comment by m-rph** [comment]
Forgot about this. I will close.

**Pull request closed without merging**